### DISCUSSION TITLE
Set-PnPUserProfileProperty : This operation requires you to be managing your own data or have administrator privileges

### DISCUSSION THREAD

**Discussion by alan-hu** [pnp/powershell#2155] [Q&A]
Found another issue #277 but my issue looks different in term of error message, so I am not sure if they are the same or different.

I have a cloud only account with active SPO Admin role, 

To reproduce, run the following:
`Connect-PnPOnline -Url https://my-tenant-admin-site.sharepoint.com -credential (get-credential)`
`Set-PnPUserProfileProperty -Account someone-else-in-my-org-upn -Property "custom-property-name" -Value "this is a single-line-of-text"`

This custom user profile property is a string type of user profile property that is editable but hidden from users' Edit Profile screen.

And we received the following error:
`Set-PnPUserProfileProperty : This operation requires you to be managing your own data or have administrator privileges.`
`At line:1 char:1`
`+ Set-PnPUserProfileProperty -Account xxx.yyy@zzz -Property ...`
`+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~`
 `   + CategoryInfo          : WriteError: (:) [Set-PnPUserProfileProperty], ServerException`
  `  + FullyQualifiedErrorId : EXCEPTION,PnP.PowerShell.Commands.UserProfiles.SetUserProfileProperty`
`

**Comment by Blue2juice** [comment]
We have just tried this and are getting the same issue and error. Tested on many of the default OOTB properties too.

**Comment by keytrap-x86** [comment]
Everybody is currently waiting for a fix from Microsoft who are aware of this issue.
Please see my comment https://github.com/pnp/powershell/issues/277#issuecomment-1350886815

**Comment by mlsc63** [comment]
Hi all, yes it looks like there is a workaround before MS patched the problem. The only way is to connect differently to the service;
Connect-PnPOnline -S -Url $url -credential $creds

**Reply by mixermandan** [reply]
This works! so weird most of the other modules fail if not using -interactive but this one needed that -S.

**Reply by augiaugi** [reply]
worked for me too :) hope to get a bug fix.

**Reply by ghost** [reply]
worked for me as well!

**Reply by AndersRask** [reply]
It is not the only way to use -SPOManagementShell and the disadvantage is that it does not work interactively.
I usually go with ACS (Legacy) App Only where you create a client id the "good old way" using the portal:
https://learn.microsoft.com/en-us/sharepoint/dev/solution-guidance/security-apponly-azureacs

But yes, piggybacking on the SPO Management Shell application works too with credentials :)

**Comment by ZzyzxIsAlreadyTaken** [comment]
I am trying this with a Azure Automation account and a Azure Runbook. Im getting the same error message. Anyone with a fix for that? :) 

Set-PnPUserProfileProperty: Line | 5 | … erprofile = Set-PnPUserProfileProperty -Account "username@tenantdomain.com … | ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ | This operation requires you to be managing your own data or have administrator privileges.

```
Connect-PnPOnline https://tenant-admin.sharepoint.com -ManagedIdentity

$setuserprofile = Set-PnPUserProfileProperty -Account "username@tenantdomain.com" -PropertyName "Initials" -Value "asdafasdaf"
```

**Reply by AndersRask** [reply]
Managed Identity will not work for this, you need to use legacy auth (SPO Management Shell or ACS) as mentioned elsewhere in this thread

**Reply by EA12** [reply]
Hi @ZzyzxIsAlreadyTaken,

@AndersRask is right.
You must navigate to your SharePoint-Admin site and open the appinv.aspx from the layouts folder.
Make sure you are global admin for this task.
https://yourTenant-admin.sharepoint.com/_layouts/15/appinv.aspx
As you already have a Runbook, I assume that you also have an App-Registration. Get the Application-ID from that App-Reg and use this information to assign App-Only Full-Control to that App-Reg.

`<AppPermissionRequests AllowAppOnlyPolicy="true">
	<AppPermissionRequest Scope="http://sharepoint/social/tenant" Right="FullControl" />
	<AppPermissionRequest Scope="http://sharepoint/content/tenant" Right="FullControl" />
</AppPermissionRequests>`

After you have added the entry and trusted it, you need create a secret in your App-Reg, if you don't already have one.
You will need the secret to connet to PnP-Online to get Set-PnPUserProfileProperty to work and avoid the error message you have got.

Make also sure that the App-Reg has proper API-Permissions. 

![image](https://github.com/pnp/powershell/assets/22025808/7e6280c6-ba9f-46cc-a5a4-19af29d28c3b)

Then connect with ClientId (which is the Application-ID of your App-Reg) and the ClientSecret of your App-Reg.

`Connect-PnPOnline -Url $adminUrl -ClientId $clientId -ClientSecret $clientSecret`

Now Set-PnPUserProfileProperty will work.

Cheers, Ronny

**Reply by Divvytown** [reply]
I am trying to follow these instructions, but when i get to the step where i need to trust the application the button is greyed out for me. I am logged in with both GA and Sharepoint Admin roles active, anyone have any tips for what i could check?

**Reply by AndersRask** [reply]
I have seen people have this issue, but haven't had it myself.
I found this old comment with 2 potential fixes, so maybe try those

_We are experiencing the same issue and have narrowed it down to the "AllowAppOnlyPolicy" flag. When set to false we can deploy an app w/o this error. – 

I have found workaround for this issue. When we get the screen with disabled button "Trust It" go to "Site Content". Find your app and click "more information" button on the app. Then click "Permissions" link. You will see the screen with your app and error message that Your tenant administrator ..bla bla bla and in the last sentence click "here" link. After that you can enter your app without problem. Hope it will help someone. –_ 

https://stackoverflow.com/questions/38504451/sharepoint-online-app-installation-error

**Reply by Divvytown** [reply]
Thanks! The suggestion to use https://tenant-admin.sharepoint.com/_layouts/15/appinv.aspx worked for me. I am now able to connect to PNP Powershell using app registration and certificate based authentication.

**Reply by lfastrup** [reply]
Thx Anders - does indeed work by registering such an App. 
I need to update some custom properties in the "good" old SharePoint user profiles. Hope we can get rid of that requirement one day. 
The App connection will work until April 2026, where ACS is retired by MS!
Anyone here knows how to go about it without ACS?

**Reply by lfastrup** [reply]
Worked it out - Create an Azure App and grant SharePoint User.ReadWrite.All plus Sites.FullControl.All
Add a self-signed certificate.

Connect as follows:
$password = (ConvertTo-SecureString -AsPlainText 'myprivatekeypassword' -Force)
Connect-PnPOnline -Url "contoso.sharepoint.com" -ClientId 6c5c98c7-e05a-4a0f-bcfa-0cfc65aa1f28 -CertificatePath 'c:\mycertificate.pfx' -CertificatePassword $password  -Tenant 'contoso.onmicrosoft.com'

I was then able to update user profile properties using the cmdlet Set-PnPUserProfileProperty

**Comment by mixermandan** [comment]
I haven’t used runbooks so I couldn’t say if you’re having an issue with passing the credentials, however the connect-pnponline command doesn’t work correctly for the set-pnpuserprofile command it’s a bug.

You must use

Connect-PnPOnline -S -Url $AdminSiteURL -Credentials $cred

**Comment by ga-cmyk** [comment]
Hi,

I am still getting this issue when using below command from PowerShell 7.  The account used for connecting has SharePoint Admin permissions. The same account is able to update the property value from UI and also from Power Automate using "/_api/SP.UserProfiles.PeopleManager/SetSingleValueProfileProperty".  So, no issues with account Permissions.

**PnP version is 3.1.0**


Connect-PnPOnline -Url $AdminSiteURL -ClientId <GUID> -Interactive
$User = "i:0#.f|membership|useremamil"
Set-PnPUserProfileProperty -Account $User -PropertyName "CustomPropertyA" -Value "ABC"

**Comment by JarvenZhang** [comment]
Also tried with CSOM by PNPcontext and exact same error. 

Set-PnPUserProfileProperty -Account xxxx -PropertyName SPS-PicturePlaceholderState -Value 1
Set-PnPUserProfileProperty: Property Not Editable: This property can only be modified by an administrator.

PS C:\Users\SPO> $peopleManager.SetSingleValueProfileProperty("i:0#.f|membership|" + "xxxxx", "SPS-PicturePlaceholderState", 0)
PS C:\Users\SPO>   $adminCtx.ExecuteQuery()
MethodInvocationException: Exception calling "ExecuteQuery" with "0" argument(s): "Property Not Editable: This property can only be modified by an administrator."
